test_runner/src/main/kotlin/ftl/json/SavedMatrix.kt

```diff
@@ -113,3 +134,12 @@ private fun TestMatrix.invalidTestOutcome() = TestOutcome(
     outcome = INVALID,
     details = invalidMatrixDetails.orUnknown()
 )
+
+private data class AppPath(
+    private val appApk: FileReference?,
+    private val testsZip: FileReference?,
+    private val appIpa: FileReference?
+) {
+    val gcsPath: String?
+        get() = (appApk ?: testsZip ?: appIpa)?.gcsPath
```

Comment:
interesting resolution, nice and neat but make sure we dont have any order of preference issues here :)

Reply:
The order doesn't matter in that case since only one will have a value that is non-null. It comes from `TestSpecification` object, and that one is a composition of specs:  
* `androidInstrumentationTest`
* `androidTestLoop`
* `androidRoboTest`
* `iosXcTest`
* `iosTestLoop`

Since we are allowed to run only one type of test per matrix only one from above is not null. All android specs have `appApk`, iOS xc -- `testsZip` and iOS loop -- `appIpa`. Let me know if that answers your question :)